python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+
+
+class TimeLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return self
+
+
+class TimestampLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> "TimestampLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.fromtimestamp(self.value / 1000000) - EPOCH).days)
+
+
+class DecimalLiteral(Literal[Decimal]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> Optional["DecimalLiteral"]:
+        if type_var.scale == abs(self.value.as_tuple().exponent):
+            return self
+        return None
+
+
+class StringLiteral(Literal[str]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(StringType)
+    def _(self, type_var: StringType) -> "StringLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return TimeLiteral(
+            int(
+                (
+                    datetime.datetime.strptime((EPOCH.strftime("%Y-%m-%d ") + self.value), "%Y-%m-%d %H:%M:%S.%f") - EPOCH
+                ).total_seconds()
+                * 1000000
+            )
+        )
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> TimestampLiteral:
+        try:
+            timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f")
+        except ValueError as e:
+            try:
+                timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f%z")
+                raise TypeError(f"Cannot convert String to Timestamp when timezone is included: {self.value}")
+            except:
+                raise
+
+        return TimestampLiteral(int((timestamp - EPOCH).total_seconds() * 1_000_000))
+
+    @to.register(TimestamptzType)
+    def _(self, type_var: TimestamptzType) -> "TimestampLiteral":
+        try:
+            timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f%z")
+        except ValueError as e:
+            try:
+                timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f")
+            except:
+                raise e
+            raise TypeError(f"Cannot convert String to Timestamptz, missing timezone: {self.value}")
+        utc_epoch = EPOCH.replace(tzinfo=pytz.UTC)
+        return TimestampLiteral(int((timestamp - utc_epoch).total_seconds() * 1_000_000))
```

Comment:
Similar to above, I think we can rewrite this:

```python
from datetime import datetime
import re

EPOCHTZ = datetime.fromisoformat("1970-01-01T00:00:00.000000+00:00")
ISO_TIMESTAMPTZ = re.compile("\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d(.\d{1,6})?[-+]\d\d(:\d\d)?")
...
    if ISO_TIMESTAMPTZ.fullmatch(self.value):
        try:
            delta = (datetime.fromisoformat(self.value) - EPOCHTZ)
            return (delta.days * 86400 + delta.seconds) * 1_000_000 + delta.microseconds
        except TypeError:
            return None
    return None
```

This differs slightly from the `timestamp` version. The regex requires an offset and it uses an epoch value that is in UTC. Also, value returned by `fromisoformat` must both have valid `tzinfo`, which is why this needs `EPOCHTZ`.

Reply:
I've also verified that this works with test cases from Java:
* -1_000_001 <=> "1969-12-31T16:59:58.999999-07:00" and "1969-12-31T23:59:58.999999+00:00"
* 1503091261919314 <-> "2017-08-18T14:21:01.919314-07:00" and "2017-08-18T14:21:01.919314+00:00"